src/hotspot/share/prims/jvmtiEventController.cpp

```diff
@@ -1207,8 +1219,20 @@ JvmtiEventController::vm_init() {
 
 void
 JvmtiEventController::vm_death() {
+  // No new event callbacks except vm_death can be called after this point.
+  AtomicAccess::store(&_execution_finished, true);
   if (JvmtiEnvBase::environments_might_exist()) {
     MutexLocker mu(JvmtiThreadState_lock);
     JvmtiEventControllerPrivate::vm_death();
   }
-}
+
+  // Some events might be still in callback for daemons threads and ServiceThread.
+  const double start = os::elapsedTime();
```

Comment:
Given we sleep for 100ms each iteration we could simply count the iterations and break at 600, rather than tracking elapsed time. Not that it makes any real difference.